zjit/src/codegen.rs

```diff
@@ -314,6 +315,43 @@ fn gen_insn(cb: &mut CodeBlock, jit: &mut JITState, asm: &mut Assembler, functio
     Some(())
 }
 
+fn gen_concatstrings(jit: &JITState, asm: &mut Assembler, strings: &[InsnId], state: &FrameState) -> Option<Opnd> {
+    // In practice, concatstrings with 0 strings doesn't seem to happen.
+    if strings.len() == 0 {
+        return None;
+    }
+
+    // Put the strings onto the stack
+    for (i, string) in strings.iter().enumerate() {
+        let dest = Opnd::mem(64, NATIVE_STACK_PTR, i32::try_from(strings.len() - i).ok()? * -SIZEOF_VALUE_I32);
+        asm.mov(dest, jit.get_opnd(*string)?);
+    }
+    // Write, then move the SP, since the source might be from the stack
+    let stack_growth = aligned_stack_bytes(strings.len());
+    let new_sp = asm.sub(NATIVE_STACK_PTR, stack_growth.into());
+    asm.mov(NATIVE_STACK_PTR, new_sp);
+
+    // Save PC since the call allocates and can raise
+    gen_save_pc(asm, state);
+
+    // We've moved the SP, so the array is at SP or one element above
```

Comment:
I was working on a branch for toregexp, which looks quite similar to this: https://github.com/ruby/ruby/compare/master...composerinteralia:ruby:zjit-toregexp?expand=1

I managed to avoid this check by growing the stack first, then putting the strings onto the stack relative to the new SP. That way the extra space for alignment ends up after the array instead of before it.

I was also playing around with a `cpush_vec`, since I knew we'd need something like this for concatstring as well (and maybe some others). Not sure about it yet though.